Sources/Purchasing/Purchases/TransactionPoster.swift

```diff
@@ -124,6 +124,40 @@ final class TransactionPoster: TransactionPosterType {
         transaction.finish(self.paymentQueueWrapper.paymentQueueWrapperType, completion: complete)
     }
 
+    static func shouldFinish(
+        transaction: StoreTransactionType,
+        for product: StoreProductType?,
+        customerInfo: CustomerInfo
+    ) -> Bool {
+        // Don't finish transactions if CustomerInfo was computed offline
+        guard !customerInfo.isComputedOffline else { return false }
+
+        // If we couldn't find the product, we can't determine if it's a consumable
+        guard let product = product else { return true }
+
+        switch product.productCategory {
+        case .subscription:
+            return true
+
+        case .nonSubscription:
+            // Only finish consumables if the server actually processed it.
+            let shouldFinish = (
+                !transaction.hasKnownTransactionIdentifier ||
+                customerInfo.nonSubscriptions.contains {
+                    $0.storeTransactionIdentifier == transaction.transactionIdentifier
+                }
+            )
```

Comment:
My head is still getting stuck here if `nonSubscription` means consumable or non-consumable and if they should be treated the same 🤷‍♂️

Reply:
Answer to this (will test to make sure), `non_subscriptions` include:
- Consumables
- Non-consumables
- Non-renewing subscriptions